kernel/frame_allocator/src/lib.rs

```diff
@@ -421,120 +648,56 @@ impl AllocatedFrames {
-        assert!(self.size_in_frames() == 1);
-        AllocatedFrame {
-            frame: *self.start(),
-            _phantom: PhantomData,
-        }
+impl<const S: FrameState> Deref for Frames<S> {
+    type Target = FrameRange;
+    fn deref(&self) -> &FrameRange {
+        &self.frames
     }
 }
-
-/// This function is a callback used to convert `UnmappedFrames` into `AllocatedFrames`.
-/// `UnmappedFrames` represents frames that have been unmapped from a page that had
-/// exclusively mapped them, indicating that no others pages have been mapped 
-/// to those same frames, and thus, they can be safely deallocated.
-/// 
-/// This exists to break the cyclic dependency cycle between this crate and
-/// the `page_table_entry` crate, since `page_table_entry` must depend on types
-/// from this crate in order to enforce safety when modifying page table entries.
-fn into_allocated_frames(frames: FrameRange) -> AllocatedFrames {
-    AllocatedFrames { frames }
+impl<const S: FrameState> Ord for Frames<S> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.frames.start().cmp(other.frames.start())
+    }
 }
-
-impl Drop for AllocatedFrames {
-    fn drop(&mut self) {
-        if self.size_in_frames() == 0 { return; }
-
-        let (list, typ) = if contains_any(&RESERVED_REGIONS.lock(), &self.frames) {
-            (&FREE_RESERVED_FRAMES_LIST, MemoryRegionType::Reserved)
-        } else {
-            (&FREE_GENERAL_FRAMES_LIST, MemoryRegionType::Free)
-        };
-        // trace!("frame_allocator: deallocating {:?}, typ {:?}", self, typ);
-
-        // Simply add the newly-deallocated chunk to the free frames list.
-        let mut locked_list = list.lock();
-        let res = locked_list.insert(Chunk {
-            typ,
-            frames: self.frames.clone(),
-        });
-        match res {
-            Ok(_inserted_free_chunk) => (),
-            Err(c) => error!("BUG: couldn't insert deallocated chunk {:?} into free frame list", c),
-        }
-        
-        // Here, we could optionally use above `_inserted_free_chunk` to merge the adjacent (contiguous) chunks
-        // before or after the newly-inserted free chunk. 
-        // However, there's no *need* to do so until we actually run out of address space or until 
-        // a requested address is in a chunk that needs to be merged.
-        // Thus, for performance, we save that for those future situations.
+impl<const S: FrameState> PartialOrd for Frames<S> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
     }
 }
-
-impl<'f> IntoIterator for &'f AllocatedFrames {
-    type IntoIter = AllocatedFramesIter<'f>;
-    type Item = AllocatedFrame<'f>;
-    fn into_iter(self) -> Self::IntoIter {
-        AllocatedFramesIter {
-            _owner: self,
-            range_iter: self.frames.iter(),
-        }
+// To Do: will this be an issue as now this applies to Chunk as well as AllocatedFrames?
```

Comment:
previously AllocatedFrames didn't have an implementation for PartialEq, only Chunk did. Will this definition of PartialEq (and the other traits) for AllocatedFrames lead to any incorrectness?

Reply:
no worries, that should be fine.